Get root namespace from the project

genlu opened the pull request:
This is the implementation for the API added in #31254. 

Also this include the change that (hopefully) clarified the "default namespace vs root namespace for VB" issue. Right now VB doesn't have the concept of "default namespace". In this change, we conjure one in workspace, and assign the value of the project's root namespace to it for now. Individual IDE features can choose to use it for their own purpose. This would enable us to introduce "default namespace" for VB project in the future. One possibility is through a new `<defaultnamespace>` msbuild property which by default set to the value of `<rootnamespace>`, but user can choose to set it to something else when `<rootnamespace>` is left empty, i.e. root namespace is `Global` and IDE can enforce namespace declaration in the project based on the value of `<defaultnamespace>`

FYI @jasonmalinowski @davkean

genlu commented on src/VisualStudio/Core/Impl/ProjectSystem/CPS/CPSProject_IWorkspaceProjectContext.cs:
```diff
@@ -123,6 +123,22 @@ public void SetOptions(string commandLineForOptions)
+
+        public void SetProperty(string name, object value)
+        {
+            if (string.Equals(name, AdditionalPropertyNames.DefaultNamespace, StringComparison.OrdinalIgnoreCase) 
```
Isn't msbuild property name case insensitive?

jasonmalinowski replied:
Possibly, but are you sure that's the right casing comparer? I'd propose that any msbuild property name normalization is done by the project system (or msbuild), and not us. We should only be getting canonical property names here.

genlu replied:
Fixed

genlu commented on src/VisualStudio/Core/Impl/ProjectSystem/CPS/CPSProject_IWorkspaceProjectContext.cs:
```diff
@@ -123,6 +123,22 @@ public void SetOptions(string commandLineForOptions)
+        public string DefaultNamespace
+        {
+            get => _visualStudioProject.DefaultNamespace;
+            private set => _visualStudioProject.DefaultNamespace = value;
```
It seems to be the pattern around here. @jasonmalinowski would know better.

jasonmalinowski replied:
CPSProject (and VisualStudioProject) is the mutable handle that provides the mutable world to the project system. So yes, this is correct.

genlu commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/CPS/IWorkspaceProjectContext.cs:
```diff
@@ -25,6 +23,9 @@ internal interface IWorkspaceProjectContext : IDisposable
         void SetOptions(string commandLineForOptions);
 
+        // Other project properties.
+        void SetProperty(string name, object value);
```
This is what we have decided. Ping @jasonmalinowski @davkean to confirm.

jasonmalinowski replied:
I went with object since we might want this for things that are also not MSBuild properties. But I'd be OK with string/string too.

genlu replied:
Changed to string/string

genlu commented on src/Workspaces/Core/Portable/Workspace/Workspace.cs:
```diff
@@ -525,6 +525,22 @@ protected internal void OnProjectNameChanged(ProjectId projectId, string name, s
+                var oldSolution = this.CurrentSolution;
+                var newSolution = this.SetCurrentSolution(oldSolution.WithProjectDefaultNamespace(projectId, defaultNamespace));
+
+                this.RaiseWorkspaceChangedEventAsync(WorkspaceChangeKind.ProjectChanged, oldSolution, newSolution, projectId);
```
I will just follow the pattern for now.

jasonmalinowski replied:
@genlu Can you at least file a bug to track the cleanup? I believe your response has been the response the last 10 times. At some point we should reconsider.

genlu replied:
In that case, i will just fix it. I thought that is a pattern we decided to follow for some reason.

genlu replied:
Fixed

jasonmalinowski replied:
Awesome, thanks!

genlu commented on src/VisualStudio/Core/Impl/ProjectSystem/CPS/CPSProject_IWorkspaceProjectContext.cs:
```diff
@@ -123,6 +123,22 @@ public void SetOptions(string commandLineForOptions)
+        public void SetProperty(string name, object value)
+        {
+            if (string.Equals(name, AdditionalPropertyNames.DefaultNamespace, StringComparison.OrdinalIgnoreCase) 
+                && _visualStudioProject.Language == LanguageNames.CSharp
```
Done

jasonmalinowski replied:
Except we also want this for VB too...please figure out how that design works here as well. :smile:

genlu replied:
Unless we want to add a `RootNamespace` property to `Project` and set that to the exact same value user can get from VB compilation options or `null` for non-VB projects, I'd stick with current logic for now.

I'm not saying the proposal above is bad, we definitely should discuss how default/root namespace should work for all roslyn workspace projects at some point. But it's not a crucial part for this change.

jasonmalinowski replied:
@genlu It's not strictly necessary, but my concern is if we're designing this into a corner here without thinking bigger picture. That, and these types are really language-agnostic and so any language checking *at all* is a very big code smell.

genlu replied:
I get what you are saying. But my gut feeling is we would end up not support "default namespace" to VB, I don't really see a way to design it so that root and default namespace could co-exist in a project without being super confusing. Plus VB users seem to be less likely to declare namespace in source (https://github.com/dotnet/vblang/issues/351), so I'm not sure there's much demand for this. Thoughts?

CyrusNajmabadi replied:
my main issue here as well is that this is hardocding C# behaviro into the common layer.  if we need c# specific behavior, it should go at teh c# level.  Or, if we want common behavior for everything but VB, then the exception shuld go in the VB layer.

For all intensive porpoises, the common layer should not reference LanguageNames.CSharp or LanguageNames.VB.  (only exception is when the common layer is truly making a component common to both and it needs to reference both names to export it.

genlu replied:
I agree with you. But we don't seem to have a language specific layer for CPS project (like what we have for legacy projects). @jasonmalinowski Any suggestions?

genlu replied:
Well, if we just set the value to whatever project system provides us, then it would make no sense for VB (not sure about F#) as the default namespace property of the project will then have a non-null value. 

One way I think we can do is, instead of having a concept of default namespace in workspace that is C# only, we change it to `RootNamespace` which just represents the value of `rootnamespace` property in project file. Then we will let each individual feature do the language check and decide how to interpreter this value.

genlu replied:
Based on offline discussion with @jasonmalinowski, now we just expose root namespace directly, and will let the individual feature decide what to do with it.

jasonmalinowski commented on src/VisualStudio/Core/Impl/ProjectSystem/CPS/AdditionalPropertyNames.cs:
```diff
@@ -0,0 +1,9 @@
+{
+    internal static class AdditionalPropertyNames
+    {
+        public const string DefaultNamespace = "rootnamespace";
```
This seems like it's deserving a of a very long comment why it's like this. :smile:

genlu replied:
Good point

genlu replied:
Just change the const name to `RootNamesapce` for the sake of clarity

genlu commented:
@jasonmalinowski Could you please take another look?

jasonmalinowski commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs:
```diff
@@ -85,6 +88,15 @@ internal VisualStudioProject(VisualStudioWorkspaceImpl workspace, HostDiagnostic
             _displayName = projectUniqueName;
             _projectUniqueName = projectUniqueName;
 
+            if (language == LanguageNames.CSharp)
```
There really shouldn't be special language checks like this. I would assume we *also* want it null for C# as a *default* as "I don't know". There are project types that will create projects like this that won't ever initialize this value, but you don't want to act upon it.

genlu replied:
That's exactly why this code wasn't in the first iteration, I had a brain fart apparently... thanks for pointing out

jasonmalinowski commented on src/VisualStudio/Core/Impl/ProjectSystem/CPS/CPSProject_IWorkspaceProjectContext.cs:
```diff
@@ -123,6 +123,24 @@ public void SetOptions(string commandLineForOptions)
+            // through `VisualBasicCompilationOptions.RootNamespace`.
+            if (name == AdditionalPropertyNames.RootNamespace && _visualStudioProject.Language == LanguageNames.CSharp)
+            {
+                DefaultNamespace = string.IsNullOrEmpty(value) ? string.Empty : value;
```
```suggestion
                DefaultNamespace = value ?? string.Empty;
```

jasonmalinowski commented on src/Workspaces/Core/Portable/Workspace/Workspace.cs:
```diff
@@ -461,207 +461,152 @@ protected virtual void CheckProjectCanBeRemoved(ProjectId projectId)
-
-                this.RaiseWorkspaceChangedEventAsync(WorkspaceChangeKind.ProjectChanged, oldSolution, newSolution, projectId);
-            }
+            this.HandleProjectChange(projectId, (id, oldSolution) => oldSolution.WithProjectOutputFilePath(id, outputFilePath));
```
It's a bit strange here that the lambda gets the ID passed through and we have both "projectId" and "id" available when they're really the same. Given we're still closing over outputFilePath (so the lambda is allocating a closure no matter what), just simplify it a bit?

genlu replied:
I figured that would be one fewer field in the closure, but that's doesn't really make much difference I guess. Fixed

jasonmalinowski commented on src/VisualStudio/Core/Impl/ProjectSystem/CPS/AdditionalPropertyNames.cs:
```diff
@@ -0,0 +1,9 @@
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.CPS
+{
+    internal static class AdditionalPropertyNames
```
Comment these are MSBuild property names?

genlu commented:
@jasonmalinowski I have addressed your comments, please take another look

genlu commented:
@jasonmalinowski So as mentioned offline, I have made the change to expose root namespace directly, instead of default namespace for C# projects. 

One thing that's not working properly is when changing the rootnamespace property value for legacy project through project property (`project -> <project name> properties... -> Default namespace (for C#)/Root namespace (for VB`)), there's no event notifying us about the change.

This commit (https://github.com/dotnet/roslyn/pull/30998/commits/10bc122de4476b0cf0d8a76f0fd4fef37ba1d44d) is an attempt to hack around this issue for legacy VB project, but I ended up reverting it because (1) I really don't like it and (2) the feature I'm working on doesn't even need it from VB project. So I will just file a bug for this and figure out if I can listen to the property change from VS.

genlu commented:
Separated the API change to another PR: #31254

Squashed all commits and fixed sync namespace refactoring to query rootnamespace instead.

Also, according to @tmeschter, there's no event we can listen to for the change to root/default namespace field in project property page in VS (see screen below), which means for legacy project system, the value will be outdated once user changed it through UI. Do you guys know if there's any way to address this issue? @CyrusNajmabadi @davkean ?

![imgo](https://user-images.githubusercontent.com/788783/48737263-7150a380-ec02-11e8-8aa2-fa7253e6daed.jpg)

genlu commented:
@jinujoseph @vatsalyaagrawal for Preview2 approval.

genlu commented:
@jasonmalinowski Could you please take another look?

genlu commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/Legacy/AbstractLegacyProject.cs:
```diff
@@ -330,13 +336,13 @@ private static void ComputeFolderNames(uint folderItemID, List<string> names, IV
+        /// This property has different meaning between C# and VB, each project type can decide how to interpret the value.
         /// </remarks>>
-        private static string GetDefaultNamespace(IVsHierarchy hierarchy, string language)
+        private static string GetRootNamespacePropertyValue(IVsHierarchy hierarchy)
```
Unfortunately, that's not the case.But I'm not too concerned about this scenario, because (1) it's not very common and (2) we are moving legacy to `IWorkspaceProjectContext`
